staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -705,7 +706,12 @@ func (c *Cacher) List(ctx context.Context, key string, resourceVersion string, p
 	}
 	filter := filterWithAttrsFunction(key, pred)
 
-	objs, readResourceVersion, err := c.watchCache.WaitUntilFreshAndList(listRV, pred.MatcherIndex(), trace)
+	matchValues := pred.MatcherIndex()
+	// TODO: Is there any better way to retrieve namespace (other than parsing "key")?
+	if namespace, ok := request.NamespaceFrom(ctx); ok && len(namespace) > 0 {
```

Comment:
@liggitt - this namespace index is something I would like to add, but I'm not super proud with this. Do you have any better ideas to get namespace here? If not, is it something that you think is acceptable?

Reply:
I definitely wouldn't recommend parsing the key. This isn't really the layer where I expected to modify the predicate based on the context... context is opaque at this layer, right? would doing this in Store#List make more sense? What is the impact if a metadata.namespace field selector is already specified? Does populating the field selector in the non-read-from-cache case force fetching of object attributes and map allocations for every item?

Reply:
> I definitely wouldn't recommend parsing the key. 

+100 - that's why I didn't go this way too.

> This isn't really the layer where I expected to modify the predicate based on the context... context is opaque at this layer, right? 

Is it? I think that we're not utilizing context enough and I would say that we should treat it completely as opaque...

> would doing this in Store#List make more sense? What is the impact if a metadata.namespace field selector is already specified? Does populating the field selector in the non-read-from-cache case force fetching of object attributes and map allocations for every item?

Those are valid questions - I will try to get back to it in couple weeks. Thanks!

Reply:
FTR (since we may want to return to that):

If we would do that in Store#List then:
- if metadata.namespace is set, that should simply be skipped (and we're already good)
- for non-cache case, if the selectors (both field and label) were empty, then it would result in fetching attributes, if any of them weren't empty, then we are fetching them anyway...
- in all cases, it would add additional processing of going over all items and ensuring they match the field selector